raster-foundry/raster-foundry
# Handle Numerical Scopes

## Overview

Prior to this PR numerical limits on scopes were not handled in any special way and were ignored. This PR does the following:
 - Updates `groundworkUser` with proper limits for project creation, project sharing, and upload size
 - Adds machinery for calculating and checking limits

### Checklist


### Notes

This takes the easier route for verifying limits in that we check if a user has reached a limit prior to continuing. This means that users _could_ upload more than 10 GB by going over the limit with an upload. However, closing this small loophole would require a significant refactor in the upload handling.

## Testing Instructions

- Set the development user to a `groundworkUser` (do this first):
```
./scripts/psql
update users set scopes = '"groundworkUser"'::jsonb
\q
```
- Make sure to restart memcached `docker-compose restart memcached`
- Assemble and start servers, get a JWT token and set it in your environment:
```
export JWT_AUTH_TOKEN={token}
```
- Attempt to create another project (should fail because the development user has more than 5 projects created)
```
echo '{"name":"Test","description":"","visibility":"PRIVATE","tileVisibility":"PRIVATE","tags":[],"isAOIProject":false}' | http POST :9100/api/projects --auth-type=jwt
```
- Create an upload with `UPLOADED` status twice - first one should work, second should fail because you're over the limit:
```
echo '{"visibility": "PRIVATE", "datasource":"b4ca6661-27fa-497e-a85a-2a5632d5fe7c","fileType":"GEOTIFF","files":["s3://rasterfoundry-development-data-us-east-1/bigtiff-cog.tif", "s3://rasterfoundry-development-data-us-east-1/large-3-band.tif", "s3://rasterfoundry-development-data-us-east-1/user-uploads/auth0|59318a9d2fbbca3e16bcfc92/19eb3b70-c7d2-49d9-aab3-277271c7a766/m_4110050_se_14_1_20140803-cog.tif"],"keepInSourceBucket":false,"layerId":"2f7173b9-63ed-41ec-a82d-dc0f67c344e3","metadata":{"acquisitionDate":"2020-01-27T19:16:31.819Z","cloudCover":0},"projectId":"48027ab1-0390-4e70-9f07-e5f12d1f085d","source":null,"uploadStatus":"UPLOADED","uploadType":"LOCAL"}' | http POST :9100/api/uploads --auth-type=jwt

echo '{"visibility": "PRIVATE", "datasource":"b4ca6661-27fa-497e-a85a-2a5632d5fe7c","fileType":"GEOTIFF","files":["s3://rasterfoundry-development-data-us-east-1/bigtiff-cog.tif", "s3://rasterfoundry-development-data-us-east-1/large-3-band.tif", "s3://rasterfoundry-development-data-us-east-1/user-uploads/auth0|59318a9d2fbbca3e16bcfc92/19eb3b70-c7d2-49d9-aab3-277271c7a766/m_4110050_se_14_1_20140803-cog.tif"],"keepInSourceBucket":false,"layerId":"2f7173b9-63ed-41ec-a82d-dc0f67c344e3","metadata":{"acquisitionDate":"2020-01-27T19:16:31.819Z","cloudCover":0},"projectId":"48027ab1-0390-4e70-9f07-e5f12d1f085d","source":null,"uploadStatus":"UPLOADED","uploadType":"LOCAL"}' | http POST :9100/api/uploads --auth-type=jwt
```
- Attempt to share a project with more than 5 people - last one should fail:
```
echo '{"objectType":"PROJECT","objectId":"7e584c31-f5d1-4a02-9428-e83006642375","subjectType":"USER","subjectId":"auth0|5b1edcf055f2302168002137","actionType":"VIEW","isActive":true}' | http POST :9100/api/projects/7e584c31-f5d1-4a02-9428-e83006642375/permissions --auth-type=jwt

echo '{"objectType":"PROJECT","objectId":"7e584c31-f5d1-4a02-9428-e83006642375","subjectType":"USER","subjectId":"default","actionType":"VIEW","isActive":true}' | http POST :9100/api/projects/7e584c31-f5d1-4a02-9428-e83006642375/permissions --auth-type=jwt

echo '{"objectType":"PROJECT","objectId":"7e584c31-f5d1-4a02-9428-e83006642375","subjectType":"USER","subjectId":"rf|airflow-user","actionType":"VIEW","isActive":true}' | http POST :9100/api/projects/7e584c31-f5d1-4a02-9428-e83006642375/permissions --auth-type=jwt

echo '{"objectType":"PROJECT","objectId":"7e584c31-f5d1-4a02-9428-e83006642375","subjectType":"USER","subjectId":"auth0|5b1ee0a0157859716f2dbd85","actionType":"VIEW","isActive":true}' | http POST :9100/api/projects/7e584c31-f5d1-4a02-9428-e83006642375/permissions --auth-type=jwt

echo '{"objectType":"PROJECT","objectId":"7e584c31-f5d1-4a02-9428-e83006642375","subjectType":"USER","subjectId":"auth0|5b1ee1ad157859716f2dbdc2","actionType":"VIEW","isActive":true}' | http POST :9100/api/projects/7e584c31-f5d1-4a02-9428-e83006642375/permissions --auth-type=jwt

echo '{"objectType":"PROJECT","objectId":"7e584c31-f5d1-4a02-9428-e83006642375","subjectType":"USER","subjectId":"auth0|5b1ee115157859716f2dbd9f","actionType":"VIEW","isActive":true}' | http POST :9100/api/projects/7e584c31-f5d1-4a02-9428-e83006642375/permissions --auth-type=jwt
```

Closes #531
